### DISCUSSION TITLE
Specifications?

### DISCUSSION THREAD

**Discussion by MarkDH102** [fhdm-dev/scoppy#21] [Q&A]
As a first delve into the Pico I'm finding this an excellent tool.
Do you have any idea of the maximum frequency that can be measured?
I have a 100kHz PWM 50/50 coming out of the Pico and scoppy sees it, but there is a pronounced curve up towards 3.3V on the rising edge rather than a nice square wave. The falling edge side is nice and square.
Just wondered if it was the pico generating the PWM or a limitation of the analogue input on the other pico running scoppy?

The 2 pico GND's are connected together and the wire carrying the PWM from one to the other is about 10cm long.
The scoppy pico is powered from the tablet and the other pico by my PC USB.

**Comment by santlaha** [comment]
Joto 500k. After that it's the limit of Pico's adc  speed. Which takes 96
cycles to capture one shot. (Using 125 Mhz clock speed )
It's possible to overclock upto 270Mhz but I haven't tried that, yet.

**Reply by MarkDH102** [reply]
OK, thanks for that.
I now have a square wave at 100kHz, self generated (not PWM). The PWM generated 100kHz also showed some strange artifacts.

```
#include <stdio.h>
#include "pico/stdlib.h"
#include "hardware/gpio.h"

int main()
{
    gpio_init(11);
    gpio_set_dir(11, GPIO_OUT);

    while (true) 
    {
		gpio_put(11, 0);
		sleep_us(5);
		gpio_put(11, 1);
		sleep_us(5);
    }

    return 0;
}
```

And I'm connected to a Galaxy TAB 5SE. I'm seeing artifacts in the signal displayed. At 250kHz, they get much worse.
I will hook a normal scope when I get the chance. To be clear a second PICO is generating the square wave, and that is all it is doing...

**Comment by fhdm-dev** [comment]
> As a first delve into the Pico I'm finding this an excellent tool.
> Do you have any idea of the maximum frequency that can be measured?
> I have a 100kHz PWM 50/50 coming out of the Pico and scoppy sees it, but there is a pronounced curve up towards 3.3V on the rising edge rather than a nice square wave. The falling edge side is nice and square.
> Just wondered if it was the pico generating the PWM or a limitation of the analogue input on the other pico running scoppy?
> 
> The 2 pico GND's are connected together and the wire carrying the PWM from one to the other is about 10cm long.
> The scoppy pico is powered from the tablet and the other pico by my PC USB.

Hi,

It would be the low sampling rate of the ADC on the Pico (running Scoppy) that's causing the issue. The Pico can only sample at a maximum of 500k samples per second and so there will only be 5 sample points per cycle when measuring your 100kHz square wave. Scoppy uses linear interpolation to join the sample points so that's not really enough points to correctly display the waveform. Sin(x)/x interpolation would improve things and this is something I might try to add in the future.

I've found that for a sine wave the upper usable limit is about 30kHz. For a square wave I guess it's going to be lower.

I'm currently looking at adding a logic analyser function into Scoppy. This will be able to sample at a much higher rate and should be better for viewing PWM signals. Not sure when it will be ready tho'.

**Comment by Gustavomurta** [comment]
The project is very interesting and has potential. I think it can improve a lot. Congratulations and thank you.

With the oscilloscope I was able to capture square frequencies up to 10 KHz with reasonable stability.

With the logical analyzer (using only one channel) I was able to capture a square signal of up to 500 KHz. Great! 

I used this ESP32 Frequency meter. 
https://www.esp32.com/viewtopic.php?t=17018

NOTE: The logical analyzer only shows 9 vertical divisions
